src/vfs2.c

```diff
@@ -2274,7 +2275,14 @@ int vfs2_add_uncommitted(sqlite3_file *file,
 	 * "early" like this is harmless and saves us from having to stash the
 	 * page numbers somewhere else in memory in between add_uncommitted and
 	 * apply, or (worse) read them back from WAL-cur. */
-	shm_add_pgno(&e->shm, e->wal_cursor, (uint32_t)frames[0].page_number);
+	uint32_t pgno = (uint32_t)frames[0].page_number;
+	shm_add_pgno(&e->shm, e->wal_cursor, pgno);
+
+	/* With every frame, we make this update. The interpretation is that
+	 * db_size would be the size of the main file in pages if we
+	 * checkpointing up to and including the current frame. Then we use the
+	 * final value to write the commit marker of the last frame. */
+	db_size = MAX(db_size, pgno);
```

Comment:
I don't get the meaning of this variable. Why is it taking the max of number of page and page number? I may misunderstanding something but how can you related those if pgno can be an update for a page that is not the last one.

Reply:
The page number is just a (one-based) index from the start of the main file. The database grows when you write to a page whose number is higher than the previous size in pages, and it never shrinks (because we don't vacuum). So the database size at any given point is the max of the page numbers from all the page writes that have happened so far.

Reply:
That makes sense, thanks! I re-read the comment having this in mind and it is perfectly clear.